### PR TITLE
Complete the helm mixin schema

### LINKED ISSUE #21: Complete mixin schema
This is a follow-up to https://github.com/deislabs/porter/pull/186. We originally added just enough support for the schema command to verify that our design for the command would work with VS Code. 

Now that the command is plumed all the way through, let's go back and complete adding all the properties, enums, etc that the helm mixin supports.

### PR DESCRIPTION
**Pull request by carolynvs**
- [x] Compliant with new plans for how to embed the entire schema doc into the porter mega manifest.
  * Use an id
  * references use the full path rooted at the id
  * pass in the wrapping action, not just the step (so that we aren't forcing mixins to maintain separate json schema documents and then merge them for the schema command).
- [x] Validates the input yaml against the json schema during `helm run`.
- [x] Write a test that validates the mixin schema against a yaml file.

Incidentally this PR also has the following changes:
* Switches from using `packr2 build` to `go generate` for consistency with the porter repo
* Deletes the generated `*-packr.go` files so we can stop dealing with them on merges/rebases.
* Moves the packr box to a field on the mixin, and ensures it is always initialized, for consistency with the porter repo.

Closes #21

### CODE REVIEW

File: pkg/helm/helpers.go [hunk 1]
```diff
@@ -23,13 +23,11 @@ func (t *testKubernetesFactory) GetClient(configPath string) (kubernetes.Interfa
 // NewTestMixin initializes a helm mixin, with the output buffered, and an in-memory file system.
 func NewTestMixin(t *testing.T) *TestMixin {
 	c := context.NewTestContext(t)
-	m := &TestMixin{
-		Mixin: &Mixin{
-			Context:       c.Context,
-			ClientFactory: &testKubernetesFactory{},
-		},
+	m := New()
```
Comment (carolynvs) [inline pkg/helm/helpers.go]: This ensures that values on the mixin struct aren't accidentally not initailized. So instead of using the struct directly, I'm using its New function.

File: pkg/helm/schema/helm.json [hunk 2]
```diff
@@ -0,0 +1,234 @@
+{
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "$id": "http://schema.porter.sh/mixins/helm.json",
```
Comment (carolynvs) [inline pkg/helm/schema/helm.json]: This is the new id pattern that I'm suggesting for our mixins

`http://schema.porter.sh/mixins/NAME.json`

eventually our builds will host the mixin schema there, to improve compatibility with most online json schema validators since they are pretty awful and even when these are relative files or just names, like "helm", it tries to resolve it to a file.

File: pkg/helm/schema/helm.json [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+  "definitions": {
+    "installStep": {
+      "type": "object",
+      "properties": {
+        "helm": {
+          "type": "object",
+          "properties": {
+            "description": {
+              "$ref": "http://schema.porter.sh/mixins/helm.json#/definitions/stepDescription"
```
Comment (carolynvs) [inline pkg/helm/schema/helm.json]: Right now it's important that mixins that use `$ref` use the fully qualified ref. Later, we can add an issue for porter to expand the ref automatically when it imports the mixin schema.

File: pkg/helm/schema/helm.json [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+            "type": "string"
+          }
+        },
+        "additionalProperties": false,
+        "required": ["name", "secret", "key"]
+      }
+    }
+  },
+  "type": "object",
+  "properties": {
```
Comment (carolynvs) [inline pkg/helm/schema/helm.json]: This is important for the other mixins. I messed up when I originally did the schema (and by default a json schema passes any document given to it, so I didn't notice). The root json object must have a type and properties. "install" etc cannot be at the root.

I've fixed that here, but I need to go back and fix porter to handle this properly when merging schema. in https://github.com/deislabs/porter/pull/212

File: pkg/helm/schema_test.go [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+		name  string
+		step  string
+		pass  bool
+		error string
+	}{
+		{"install", "testdata/install-input.yaml", true, ""},
+		{"status", "testdata/status-input.yaml", true, ""},
+		{"upgrade", "testdata/upgrade-input.yaml", true, ""},
+		{"uninstall", "testdata/status-input.yaml", true, ""},
+		{"install.missing-desc", "testdata/bad-install-input.missing-desc.yaml", false, "install.0.helm.description: String length must be greater than or equal to 1"},
```
Comment (carolynvs) [inline pkg/helm/schema_test.go]: These negative tests are extremely important for json schema validation. Mostly to validate that our schema actually works because schema _passes by default_. I need another pass once we are sure this is all correct to fill in a slew of failure edge case tests for every single property if we really want to know if they are working as we expect.

File: pkg/helm/step.go [hunk 6]
```diff
@@ -2,7 +2,7 @@ package helm
 
 type Step struct {
 	Description string       `yaml:"description"`
-	Outputs     []HelmOutput `yaml:"outputs"`
+	Outputs     []HelmOutput `yaml:"outputs,omitempty"`
```
Comment (carolynvs) [inline pkg/helm/step.go]: This `omitempty` ensures that when we have tests that go from a go struct to yaml back to json to validate things (exhausting I know, but we have tests like that already), that we don't serialize an empty list of outputs and trigger more of the json validation.

File: pkg/helm/testdata/bad-install-input.missing-desc.yaml [hunk 7]
```diff
@@ -0,0 +1,5 @@
+install:
+- helm:
+    name: wordpress
+    chart: stable/wordpress
+    description: ""
```
Comment (carolynvs) [inline pkg/helm/testdata/bad-install-input.missing-desc.yaml]: I have a validation on description that ensures that it is at least 1 char long.

**Comment by carolynvs** [comment]
Once this is reviewed and accepted, I'll write up a document on the docs site explaining how to write a schema document for a mixin using the explanatory comments in the review as a guide.

File: pkg/helm/schema_test.go (hunk 5 above)
Comment (jeremyrickard) [reply pkg/helm/schema_test.go]: _schema passes by default_

Computers were a mistake.

File: .gitignore [hunk 8]
```diff
@@ -1,2 +1,3 @@
 /bin
 /helm
+*-packr.go
```
Comment (jeremyrickard) [inline .gitignore]: 🍶
